span_sinks.go

```diff
@@ -0,0 +1,350 @@
+		} else {
+			log.WithFields(logrus.Fields{
+				"traces":        len(finalTraces),
+				logrus.ErrorKey: err}).Warn("Error flushing traces to Datadog")
+		}
+		for service, count := range serviceCount {
+			dd.stats.Count(totalSpansFlushedMetricKey, count, []string{"sink:datadog", fmt.Sprintf("service:%s", service)}, 1)
+		}
+	} else {
+		log.Info("No traces to flush to Datadog, skipping.")
+	}
+}
+
+// lightStepSpanSink is a sink for spans to be sent to the LightStep client.
+type lightStepSpanSink struct {
+	tracer       opentracing.Tracer
+	stats        *statsd.Client
+	commonTags   map[string]string
+	mutex        *sync.Mutex
+	serviceCount map[string]int64
+}
+
+// NewLightStepSpanSink creates a new instance of a LightStepSpanSink.
+func NewLightStepSpanSink(config *Config, stats *statsd.Client, commonTags map[string]string) (*lightStepSpanSink, error) {
+
+	var host *url.URL
+	host, err := url.Parse(config.TraceLightstepCollectorHost)
+	if err != nil {
+		log.WithError(err).WithField(
+			"host", config.TraceLightstepCollectorHost,
+		).Error("Error parsing LightStep collector URL")
+		return &lightStepSpanSink{}, err
+	}
+
+	port, err := strconv.Atoi(host.Port())
+	if err != nil {
+		port = lightstepDefaultPort
+	} else {
+		log.WithError(err).WithFields(logrus.Fields{
+			"port":         port,
+			"default_port": lightstepDefaultPort,
+		}).Warn("Error parsing LightStep port, using default")
+	}
+
+	reconPeriod, err := time.ParseDuration(config.TraceLightstepReconnectPeriod)
+	if err != nil {
+		log.WithError(err).WithFields(logrus.Fields{
+			"interval":         config.TraceLightstepReconnectPeriod,
+			"default_interval": lightstepDefaultInterval,
+		}).Warn("Failed to parse reconnect duration, using default.")
+		reconPeriod = lightstepDefaultInterval
+	}
+
+	log.WithFields(logrus.Fields{
+		"Host": host.Hostname(),
+		"Port": port,
+	}).Info("Dialing lightstep host")
+
+	maxSpans := config.TraceLightstepMaximumSpans
+	if maxSpans == 0 {
+		maxSpans = config.SsfBufferSize
+		log.WithField("max spans", maxSpans).Info("Using default maximum spans — ssf_buffer_size — for LightStep")
+	}
+
+	lightstepTracer := lightstep.NewTracer(lightstep.Options{
+		AccessToken:     config.TraceLightstepAccessToken,
+		ReconnectPeriod: reconPeriod,
+		Collector: lightstep.Endpoint{
+			Host:      host.Hostname(),
+			Port:      port,
+			Plaintext: true,
+		},
+		UseGRPC:          true,
+		MaxBufferedSpans: maxSpans,
+	})
+
+	return &lightStepSpanSink{
+		tracer:       lightstepTracer,
+		stats:        stats,
+		serviceCount: make(map[string]int64),
+		mutex:        &sync.Mutex{},
+	}, nil
+}
+
+// Name returns this sink's name.
+func (ls *lightStepSpanSink) Name() string {
+	return "lightstep"
+}
+
+// Ingest takes in a span and passed it along to the LS client after
+// some sanity checks and improvements are made.
+func (ls *lightStepSpanSink) Ingest(ssfSpan ssf.SSFSpan) error {
+	ls.mutex.Lock()
+	defer ls.mutex.Unlock()
```

Comment:
Sure, but what I'm asking is - that's only used at one point way down below. Is it possible for us to scope the mutex just around the point where we increment that, or do we need to take it out at the beginning of the function?